### ISSUE TITLE
Add sensuctl create command to the CLI

### ISSUE THREAD

**Issue by echlebek** [sensu/sensu-go#1056]
`sensuctl` should be able to create resources from files that specify the resource types to be created.

The UI should look something like
```
sensuctl create -f big_list_o_checks.json
```

To accomplish this, we'll need to add a wrapper type that can specify any number of any resources, and a parser for the type to translate it into Sensu resources.

After parsing, the command will make one or more HTTP requests to the backend to create whatever resources are spec'd by the input file.

The command should only make HTTP requests once all resources are validated. If there are errors in the resource definition, one or more _helpful_ errors messages should be output, detailing the problems. In the context of this tool, _helpful_ means that a person could reasonably use the error message to diagnose the problem with their input spec.

If the `-f` flag is omitted, then `sensuctl create` should expect input from stdin.

**Comment by calebhailey** [comment]
Nice!

**Comment by jamesdphillips** [comment]
I do wonder if we should add a new endpoint for these bulk operations, say `PATCH /`. One problem the current legacy importer has is that if one HTTP request fails, the command exits and there is a no rollback. If you were importing a substantial amount of resources, you end up in a state where it's no easy to determine with operations succeeded and which resources have yet to be imported.

I suspect it will also be easier to merge resources on the backend than in the CLI.

**Comment by grepory** [comment]
I wish there was a standard pseudocode, but:

```ruby
for check in checks do
  err = api_create_check(check)
  if err != None then
    puts "error creating check id=#{check.id}: #{err}"
  end
end
```

And either fail at that error or continue.

I do like the idea of PATCH, though, because we could re-run create -f without throwing "resource already exists" errors or skipping stuff because the resource already exists. And people would have the opportunity to update.

Thoughts?

**Comment by echlebek** [comment]
@jamesdphillips,

Failed HTTP requests should probably be retried. In terms of dealing with rollback, perhaps we can get by with using PUT semantics so that resources can simply be replaced if the overall command fails. Let's look into this some more - we might need to implement something transactional if that's not going to work.

I'd like merging to not depend on the backend - it should be essentially a pure function that doesn't require any communication.

**Comment by echlebek** [comment]
I'm not keen on having `PATCH /`, because the root resource doesn't mean all resources.

**Comment by grepory** [comment]
@echlebek oh definitely not PATCH /, but PATCH for individual resources.

I don't mean failed HTTP requests, but failed creates because of conversion errors of some kind.

**Comment by echlebek** [comment]
@grepory I think we can guarantee correct data structures through Validate, no?

**Comment by grepory** [comment]
What about sensuctl / sensu-backend mismatches?

**Comment by echlebek** [comment]
I think in that case, bailing with `400 Bad Request` is quite reasonable, maybe with a hint that there is a version mismatch.

**Comment by jamesdphillips** [comment]
> Failed HTTP requests should probably be retried. In terms of dealing with rollback, perhaps we can get by with using PUT semantics so that resources can simply be replaced if the overall command fails. 

Falling back to a PUT looks like a cudgel when compared to being able to bundle all our operations into a single etcd transaction.

> I'd like merging to not depend on the backend - it should be essentially a pure function that doesn't require any communication.

No disagreement that the merge should simply be a pure function, however, my immediate thought was more to do with having more latency in the CLI than on the backend. Being able to retrieve the resource, merge and persist all in one seems more ideal to me. 🤷‍♂️

**Comment by jamesdphillips** [comment]
Re versioning, I think @amdprophet had a ticket from a while ago https://github.com/sensu/sensu-go/issues/100.

**Comment by echlebek** [comment]
@jamesdphillips yeah, I think I see where you are coming from with that. My immediate assumption was that an import command would only handle merging fragments that were passed to it - we wouldn't attempt to merge anything with existing Sensu data structures. @portertech can you comment on that?

**Comment by grepory** [comment]
I want to decrease the scope of this ticket and file additional tickets to address some of the concerns here.

- Let's break out a ticket for version checking in the API / sensuctl.
- sensuctl create -f does not need to update everything in a single etcd transaction (for now) as that's a much bigger body of work to allow that and I think PUT / is an API antipattern.

Individual resources should be validated with the CLI. I think if we tried to validate the whole file, with some helpful output, before attempting to hit the API that it'd be a good user experience. E.g.

sensuctl create -v -f some_file.json
Validating check 'check-disk': ok
Validating handler 'slack': ok
Validating mutator 'graphite': error: <handler.Validate() error message goes here>

With out -v (for verbose), we'd only do error messages.

create -f defaults to the current org/env unless they have specified org/env in the object.

We need to make sure that sensuctl can create outside of the currently configured org/env. I believe that work has already been done, if not file a ticket.

After the whole file passes validation, we attempt to hit the API. This is where version checking, once implemented takes place, but until then we should just begin the API requests. Version checking can be scoped out later.

**Comment by thomasriley** [comment]
The ability to import & update configuration within Sensu via sensuctl in this way will be super useful for the work going on within the Sensu community for writing the new Configuration Management code for Sensu-Go.

**Comment by echlebek** [comment]
This is blocked by #1153

**Comment by portertech** [comment]
This is no longer blocked by https://github.com/sensu/sensu-go/issues/1153 (Closed by #1154).

This work should take priority over the other milestone items, as it's blocking CM work (Puppet Module, Chef Cookbook, etc) and is the foundation for the new approach to 1.x -> 2.x configuration migration.
